[RFC-0003] Implement OCIRepository reconciliation (#788)
fluxcd/source-controller: docs/spec/v1beta2/ocirepositories.md

# OCI Repositories

The `OCIRepository` API defines a Source to produce an Artifact for an OCI
repository.

## Example

> The following is an example of a OCIRepository. It creates a tarball

Reviewer: ```suggestion
The following is an example of an OCIRepository. It creates a tarball
```

     Observed Generation:     1
     URL:                     http://source-controller.source-system.svc.cluster.local./gitrepository/default/podinfo/latest.tar.gz
   Events:
     Type    Reason               Age   From               Message
     ----    ------               ----  ----               -------
     Normal  NewArtifact          62s   source-controller  stored artifact with digest '3b6cdcc7adcc9a84d3214ee1c029543789d90b5ae69debe9efa3f66e982875de' from 'oci://ghcr.io/stefanprodan/manifests/podinfo'
   ```

## Writing an OCIRepository spec

As with all other Kubernetes config, a OCIRepository needs `apiVersion`,
`kind`, and `metadata` fields. The name of a OCIRepository object must be a
valid [DNS subdomain name](https://kubernetes.io/docs/concepts/overview/working-with-objects/names#dns-subdomain-names).

A OCIRepository also needs a
[`.spec` section](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status).

### URL

`.spec.url` is a required field that specifies the address of the
container image repository in the format `oci://<host>:<port>/<org-name>/<repo-name>`.

> **Note:** that specifying a tag or digest is not in accepted for this field.

Reviewer: Maybe? 

```suggestion
**Note:** that specifying a tag or digest is not acceptable for this field.
```

### Provider

`.spec.provider` is an optional field that allows specifying an OIDC provider used for
authentication purposes.

Supported options are:

- `generic`
- `aws`
- `azure`
- `gcp`

The `generic` provider can be used for public repositories or when
static credentials are used for authentication, either with
`spec.secretRef` or `spec.serviceAccountName`.
If you do not specify `.spec.provider`, it defaults to `generic`.

The `aws` provider can be used when the source-controller service account

> is associate with an AWS IAM Role using IRSA that grants read-only access to ECR.

Reviewer: ```suggestion
is associated with an AWS IAM Role using IRSA that grants read-only access to ECR.
```

> The `azure` provider can be used when the source-controller pods are associate

Reviewer: ```suggestion
The `azure` provider can be used when the source-controller pods are associated
```

with an Azure AAD Pod Identity that grants read-only access to ACR.

The `gcp` provider can be used when the source-controller service account

> is associate with a GCP IAM Role using Workload Identity that grants

Reviewer: ```suggestion
is associated with a GCP IAM Role using Workload Identity that grants
```

read-only access to Artifact Registry.

### Secret reference

`.spec.secretRef.name` is an optional field to specify a name reference to a
Secret in the same namespace as the OCIRepository, containing authentication
credentials for the OCI repository.

> This secret is expected to be in the same format as for[`imagePullSecrets`][image-pull-secrets].

Reviewer: ```suggestion
This secret is expected to be in the same format as [`imagePullSecrets`][image-pull-secrets].
```

exclusions](#sourceignore-file). See [excluding files](#excluding-files)
for more information.

### Suspend

`.spec.suspend` is an optional field to suspend the reconciliation of a
OCIRepository. When set to `true`, the controller will stop reconciling the
OCIRepository, and changes to the resource or in the OCI repository will not
result in a new Artifact. When the field is set to `false` or removed, it will
resume.

## Working with OCIRepositories

### Excluding files

By default, files which match the [default exclusion rules](#default-exclusions)
are excluded while archiving the OCI repository contents as an Artifact.
It is possible to overwrite and/or overrule the default exclusions using
the [`.spec.ignore` field](#ignore).

```yaml
---
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: OCIRepository
metadata:
  name: <repository-name>
spec:
  ignore: |
    # exclude all
    /*
    # include deploy dir
    !/deploy
    # exclude file extensions from deploy dir
    /deploy/**/*.md
    /deploy/**/*.txt
```

### Triggering a reconcile

To manually tell the source-controller to reconcile a OCIRepository outside the

> [specified interval window](#interval), a OCIRepository can be annotated with

Reviewer: ```suggestion
[specified interval window](#interval), an OCIRepository can be annotated with
```

`reconcile.fluxcd.io/requestedAt: <arbitrary value>`. Annotating the resource
queues the OCIRepository for reconciliation if the `<arbitrary-value>` differs
from the last value the controller acted on, as reported in
[`.status.lastHandledReconcileAt`](#last-handled-reconcile-at).

Using `kubectl`:

```sh
kubectl annotate --field-manager=flux-client-side-apply --overwrite ocirepository/<repository-name> reconcile.fluxcd.io/requestedAt="$(date +%s)"
```

Using `flux`:

```sh
flux reconcile source oci <repository-name>
```

### Waiting for `Ready`

When a change is applied, it is possible to wait for the OCIRepository to reach
a [ready state](#ready-gitrepository) using `kubectl`:

```sh
kubectl wait gitrepository/<repository-name> --for=condition=ready --timeout=1m
```

### Suspending and resuming

When you find yourself in a situation where you temporarily want to pause the

> reconciliation of a OCIRepository, you can suspend it using the

Reviewer: ```suggestion
reconciliation of an OCIRepository, you can suspend it using the
```

[`.spec.suspend` field](#suspend).

#### Suspend an OCIRepository

In your YAML declaration:

```yaml
---
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: OCIRepository
metadata:
  name: <repository-name>
spec:
  suspend: true
```

Using `kubectl`:

```sh
kubectl patch ocirepository <repository-name> --field-manager=flux-client-side-apply -p '{\"spec\": {\"suspend\" : true }}'
```

Using `flux`:

```sh
flux suspend source oci <repository-name>
```

> **Note:** When a OCIRepository has an Artifact and is suspended, and this

Reviewer: ```suggestion
**Note:** When an OCIRepository has an Artifact and it is suspended, and this
```

can be retrieved in-cluster from the `.status.artifact.url` HTTP address.

#### Artifact example

```yaml
---
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: OCIRepository
metadata:
  name: <repository-name>
status:
  artifact:
    checksum: e750c7a46724acaef8f8aa926259af30bbd9face2ae065ae8896ba5ee5ab832b
    lastUpdateTime: "2022-06-29T06:59:23Z"
    path: ocirepository/<namespace>/<repository-name>/<digest>.tar.gz
    revision: master/363a6a8fe6a7f13e05d34c163b0ef02a777da20a
    url: http://source-controller.<namespace>.svc.cluster.local./ocirepository/<namespace>/<repository-name>/<digest>.tar.gz
```

#### Default exclusions

The following files and extensions are excluded from the Artifact by
default:

- Git files (`.git/, .gitignore, .gitmodules, .gitattributes`)
- File extensions (`.jpg, .jpeg, .gif, .png, .wmv, .flv, .tar.gz, .zip`)
- CI configs (`.github/, .circleci/, .travis.yml, .gitlab-ci.yml, appveyor.yml, .drone.yml, cloudbuild.yaml, codeship-services.yml, codeship-steps.yml`)
- CLI configs (`.goreleaser.yml, .sops.yaml`)
- Flux v1 config (`.flux.yaml`)

To define your own exclusion rules, see [excluding files](#excluding-files).

### Conditions

> A OCIRepository enters various states during its lifecycle, reflected as

Reviewer: ```suggestion
OCIRepository has various states during its lifecycle, reflected as
```

[Kubernetes Conditions][typical-status-properties].
It can be [reconciling](#reconciling-ocirepository) while fetching the remote
state, it can be [ready](#ready-ocirepository), or it can [fail during
reconciliation](#failed-ocirepository).

The OCIRepository API is compatible with the [kstatus specification][kstatus-spec],
and reports `Reconciling` and `Stalled` conditions where applicable to
provide better (timeout) support to solutions polling the OCIRepository to
become `Ready`.

#### Reconciling OCIRepository

> The source-controller marks a OCIRepository as _reconciling_ when one of the

Reviewer: ```suggestion
The source-controller marks an OCIRepository as _reconciling_ when one of the
```

following is true:

- There is no current Artifact for the OCIRepository, or the reported Artifact
  is determined to have disappeared from the storage.
- The generation of the OCIRepository is newer than the [Observed
  Generation](#observed-generation).
- The newly resolved Artifact digest differs from the current Artifact.

When the OCIRepository is "reconciling", the `Ready` Condition status becomes
`False`, and the controller adds a Condition with the following attributes to
the OCIRepository's `.status.conditions`:

- `type: Reconciling`
- `status: "True"`
- `reason: NewGeneration` | `reason: NoArtifact` | `reason: NewRevision`

If the reconciling state is due to a new revision, an additional Condition is
added with the following attributes:

- `type: ArtifactOutdated`
- `status: "True"`
- `reason: NewRevision`

Both Conditions have a ["negative polarity"][typical-status-properties],
and are only present on the OCIRepository while their status value is `"True"`.

#### Ready OCIRepository

> The source-controller marks a OCIRepository as _ready_ when it has the

Reviewer: ```suggestion
The source-controller marks an OCIRepository as _ready_ when it has the
```
